### PR TITLE
INTMDB-301: Feature add: Add support for management of federated authentication configuration

### PR DESCRIPTION
**Pull request by martinstibbe**
## Description

Add support for managment of federated authentication configuration

Link to any related issue(s): 
https://jira.mongodb.org/browse/INTMDB-301

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added any necessary documentation (if appropriate)
- [x] I have run `make fmt` and formatted my code

### CODE REVIEW

**Review by andreaangiolillo — COMMENTED** [review]
Thanks for adding these resources. I have some minor comments

**Comment by colm-quinn** [comment]
Tagging @blva @tibulca @fmenezes, can one of you review these Terraform PR's, following the guidance Andrea added to the wiki? Thanks!

File: mongodbatlas/federation_settings.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	"fmt"
+	"net/http"
+)
+
+const federationSettingsBasePath = "api/atlas/v1.0/orgs/%s/federationSettings"
+const federationSettingsDeleteBasePath = "api/atlas/v1.0/federationSettings"
+
+// FederatedSettingsService is an interface for working with the Federation Settings
+// endpoints of the MongoDB Atlas API.
+// See more: https://www.mongodb.com/docs/atlas/reference/api/federation-configuration/
```
[the remark this answers is not in the document]
Comment (gssbzn) [reply mongodbatlas/federation_settings.go:28]: missing comment, for godocs, a new line doesn't mean a new line when rendered you need the space so they render better, see https://pkg.go.dev/go.mongodb.org/atlas@v0.16.0/mongodbatlas#AccessListAPIKeysService as an example of how it renders when adding the new line

File: mongodbatlas/federation_settings.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	ID                     string   `json:"id,omitempty"`
+	IdentityProviderID     string   `json:"identityProviderId,omitempty"`
+	IdentityProviderStatus string   `json:"identityProviderStatus,omitempty"`
+}
+
+// Get gets Federated Settings for an organization.
+//
+// See more: https://www.mongodb.com/docs/atlas/reference/api/org-get-federation-settings/#std-label-atlas-org-get-federation-settings/
+
+func (s *FederatedSettingsServiceOp) Get(ctx context.Context, orgID string) (*FederatedSettings, *Response, error) {
```
[the remark this answers is not in the document]
Comment (gssbzn) [reply mongodbatlas/federation_settings.go:55]: missing comment here, extra empty line between docs and method

File: mongodbatlas/federation_settings.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+const federationSettingsBasePath = "api/atlas/v1.0/orgs/%s/federationSettings"
+const federationSettingsDeleteBasePath = "api/atlas/v1.0/federationSettings"
+
+// FederatedSettingsService is an interface for working with the Federation Settings
+// endpoints of the MongoDB Atlas API.
+// See more: https://www.mongodb.com/docs/atlas/reference/api/federation-configuration/
+type FederatedSettingsService interface {
+	Get(context.Context, string) (*FederatedSettings, *Response, error)
+	Delete(context.Context, string) (*Response, error)
+}
```
Comment (gssbzn) [inline mongodbatlas/federation_settings.go:32]: Big change, we organize the client the same way we orgnaize the docs
From [README](https://github.com/mongodb/go-client-mongodb-atlas#usage)
>The services of a client divide the API into logical chunks and correspond to the structure of the Atlas API documentation at https://docs.atlas.mongodb.com/api/.

This means this should be 
```suggestion
type FederatedAuthenticationConfigurationService interface {
	Get(context.Context, string) (*FederatedSettings, *Response, error)
	ListConnectedOrgs(...)
	GetOneConnectedOrg(...)
	UpdateConnectedOrg(...)
	RemoveConnectedOrg(...)
	ListRoleMapping(...)
	CreateRoleMapping(...)
	// And the rest following the docs
	Delete(context.Context, string) (*Response, error)
}
```

Comment (gssbzn) [reply mongodbatlas/federation_settings.go:32]: you can keep the different files for the different sections if you want but there should be a single service all the files reference

**Review by gssbzn — CHANGES_REQUESTED** [review]
There's a big change required here which is orgnaizing all the added method under the same service to follow our practice of grouping the same way the docs group

File: mongodbatlas/federation_settings.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+// FederatedSettingsService is an interface for working with the Federation Settings
+// endpoints of the MongoDB Atlas API.
+// See more: https://www.mongodb.com/docs/atlas/reference/api/federation-configuration/
+type FederatedSettingsService interface {
+	Get(context.Context, string) (*FederatedSettings, *Response, error)
+	Delete(context.Context, string) (*Response, error)
+	ListConnectedOrgs(context.Context, *ListOptions, string) (*FederatedSettingsConnectedOrganizations, *Response, error)
+	GetConnectedOrgs(context.Context, string, string) (*FederatedSettingsConnectedOrganization, *Response, error)
+	UpdateConnectedOrgs(context.Context, string, string, *FederatedSettingsConnectedOrganization) (*FederatedSettingsConnectedOrganization, *Response, error)
```
Comment (gssbzn) [inline mongodbatlas/federation_settings.go:34]:
```suggestion
	UpdateConnectedOrg(context.Context, string, string, *FederatedSettingsConnectedOrganization) (*FederatedSettingsConnectedOrganization, *Response, error)
```

File: mongodbatlas/federation_settings.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	DeleteConnectedOrgs(context.Context, string, string) (*Response, error)
```
Comment (gssbzn) [inline mongodbatlas/federation_settings.go:35]:
```suggestion
	DeleteConnectedOrg(context.Context, string, string) (*Response, error)
```

File: mongodbatlas/federation_settings.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+	ListRoleMapping(context.Context, *ListOptions, string, string) (*FederatedSettingsOrganizationRoleMappings, *Response, error)
```
Comment (gssbzn) [inline mongodbatlas/federation_settings.go:37]:
```suggestion
	ListRoleMappings(context.Context, *ListOptions, string, string) (*FederatedSettingsOrganizationRoleMappings, *Response, error)
```

File: mongodbatlas/federation_settings.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+	GetRoleMapping(context.Context, string, string, string) (*FederatedSettingsOrganizationRoleMapping, *Response, error)
+	CreateRoleMapping(context.Context, string, string, *FederatedSettingsOrganizationRoleMapping) (*FederatedSettingsOrganizationRoleMapping, *Response, error)
+	UpdateRoleMapping(context.Context, string, string, string, *FederatedSettingsOrganizationRoleMapping) (*FederatedSettingsOrganizationRoleMapping, *Response, error)
+	DeleteRoleMapping(context.Context, string, string, string) (*Response, error)
+
+	ListIdentityProvider(context.Context, *ListOptions, string) ([]FederatedSettingsIdentityProvider, *Response, error)
```
Comment (gssbzn) [inline mongodbatlas/federation_settings.go:43]:
```suggestion
	ListIdentityProviders(context.Context, *ListOptions, string) ([]FederatedSettingsIdentityProvider, *Response, error)
```

File: mongodbatlas/federation_settings_connected_org_config.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+type RoleMappings struct {
+	ExternalGroupName string             `json:"externalGroupName,omitempty"`
+	ID                string             `json:"id,omitempty"`
+	RoleAssignments   []*RoleAssignments `json:"roleAssignments,omitempty"`
+}
+
+// List gets all Federated Settings Connected Organization (Org-Mappings).
+//
+// See more: https://www.mongodb.com/docs/atlas/reference/api/org-mappings-return-all/
+func (s *FederatedSettingsServiceOp) ListConnectedOrgs(ctx context.Context, opts *ListOptions, federationSettingsID string) (*FederatedSettingsConnectedOrganizations, *Response, error) {
```
Comment (gssbzn) [inline mongodbatlas/federation_settings_connected_org_config.go:53]: let's please keep list options always last 
```suggestion
func (s *FederatedSettingsServiceOp) ListConnectedOrgs(ctx context.Context, federationSettingsID string, opts *ListOptions,) (*FederatedSettingsConnectedOrganizations, *Response, error) {
```

**Review by gssbzn — COMMENTED** [review]
Some small nits, I still see many of andreas comments unaddressed, common themes, godocs need to be reviewed, and let's keep list options as last params

File: mongodbatlas/federation_settings_organization_role_mapping.go [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+	GroupID string `json:"groupId,omitempty"`
+	OrgID   string `json:"orgId,omitempty"`
+	Role    string `json:"role,omitempty"`
+}
+
+// List gets all Federated Settings Identity Providers for an organization.
+//
+// See more: https://www.mongodb.com/docs/atlas/reference/api/role-mapping-return-all/
+func (s *FederatedSettingsServiceOp) ListRoleMapping(ctx context.Context, opts *ListOptions, federationSettingsID, orgID string) (*FederatedSettingsOrganizationRoleMappings, *Response, error) {
+	basePath := fmt.Sprintf(federationSettingsOrganizationRoleMappingBasePath, federationSettingsID, orgID)
```
[the remark this answers is not in the document]
Comment (martinstibbe) [reply mongodbatlas/federation_settings_organization_role_mapping.go:58]: Added

File: mongodbatlas/federation_settings_organization_role_mapping.go [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+type RoleAssignments struct {
+	GroupID string `json:"groupId,omitempty"`
+	OrgID   string `json:"orgId,omitempty"`
+	Role    string `json:"role,omitempty"`
+}
+
+// List gets all Federated Settings Identity Providers for an organization.
+//
+// See more: https://www.mongodb.com/docs/atlas/reference/api/role-mapping-return-all/
+func (s *FederatedSettingsServiceOp) ListRoleMapping(ctx context.Context, opts *ListOptions, federationSettingsID, orgID string) (*FederatedSettingsOrganizationRoleMappings, *Response, error) {
```
[the remark this answers is not in the document]
Comment (martinstibbe) [reply mongodbatlas/federation_settings_organization_role_mapping.go]: Refactored

File: mongodbatlas/federation_settings_connected_org_config.go (hunk 8 above)
Comment (martinstibbe) [reply mongodbatlas/federation_settings_connected_org_config.go]: Refactored

File: mongodbatlas/federation_settings.go [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+	path := fmt.Sprintf("%s/%s", basePath, federationSettingsID)
+
+	req, err := s.Client.NewRequest(ctx, http.MethodDelete, path, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := s.Client.Do(ctx, req, nil)
+
+	return resp, err
```
[the remark this answers is not in the document]
Comment (martinstibbe) [reply mongodbatlas/federation_settings.go]: Refactored

File: mongodbatlas/federation_settings_connected_org_config.go [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+
+// FederatedSettingsIdentityProviderServiceOp handles communication with the FederatedSettings Connected Organization related methods of the
+// MongoDB Atlas API.
+type FederatedSettingsConnectedOrganizationSeviceOp service
+
+var _ FederatedSettingsConnectedOrganizationService = &FederatedSettingsConnectedOrganizationSeviceOp{}
+
+// A Resource describes a specific resource the Role will allow operating on.
+
+// FederatedSettings represents a FederatedSettings Connected Organization.
```
[the remark this answers is not in the document]
Comment (gssbzn) [reply mongodbatlas/federation_settings_connected_org_config.go:27]: missing comment

File: mongodbatlas/federation_settings_connected_org_config.go [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+	UserConflicts            *UserConflicts  `json:"userConflicts,omitempty"`
+}
+
+type RoleMappings struct {
+	ExternalGroupName string             `json:"externalGroupName,omitempty"`
+	ID                string             `json:"id,omitempty"`
+	RoleAssignments   []*RoleAssignments `json:"roleAssignments,omitempty"`
+}
+
+// List gets all Federated Settings Connected Organization (Org-Mappings).
```
Comment (gssbzn) [inline mongodbatlas/federation_settings_connected_org_config.go:50]:
```suggestion
// ListConnectedOrgs gets all Federated Settings Connected Organization (Org-Mappings).
```

File: mongodbatlas/federation_settings_connected_org_config.go [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+	}
+
+	if l := root.Links; l != nil {
+		resp.Links = l
+	}
+
+	return root, resp, nil
+}
+
+// Get getsFederated Settings Connected Organization (Org-Mapping).
```
Comment (gssbzn) [inline mongodbatlas/federation_settings_connected_org_config.go:82]:
```suggestion
// GetConnectedOrg getsFederated Settings Connected Organization (Org-Mapping).
```

File: mongodbatlas/federation_settings_connected_org_config.go [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+	root := new(FederatedSettingsConnectedOrganization)
+	resp, err := s.Client.Do(ctx, req, root)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return root, resp, err
+}
+
+// Update updates Federated Settings Connected Organization (Org-Mapping).
```
Comment (gssbzn) [inline mongodbatlas/federation_settings_connected_org_config.go:111]:
```suggestion
// UpdateConnectedOrg updates Federated Settings Connected Organization (Org-Mapping).
```

File: mongodbatlas/federation_settings_connected_org_config.go [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+	root := new(FederatedSettingsConnectedOrganization)
+	resp, err := s.Client.Do(ctx, req, root)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return root, resp, err
+}
+
+// Delete deletes federation setting for org mapping.
```
Comment (gssbzn) [inline mongodbatlas/federation_settings_connected_org_config.go:144]:
```suggestion
//  DeleteConnectedOrg deletes federation setting for org mapping.
```

File: mongodbatlas/federation_settings_identity_provider.go [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+
+// FederatedSettingsIdentityProviderServiceOp handles communication with the FederatedSettings related methods of the
+// MongoDB Atlas API.
+type FederatedSettingsIdentityProviderServiceOp service
+
+var _ FederatedSettingsIdentityProviderService = &FederatedSettingsIdentityProviderServiceOp{}
+
+// A Resource describes a specific resource the Role will allow operating on.
+
+// FederatedSettings represents a FederatedSettings List.
```
[the remark this answers is not in the document]
Comment (gssbzn) [reply mongodbatlas/federation_settings_identity_provider.go:28]: missing comment

File: mongodbatlas/federation_settings.go (hunk 7 above)
Comment (martinstibbe) [reply mongodbatlas/federation_settings.go]: Refactored to ListIdentityProviders(context.Context, string, *ListOptions) ([]FederatedSettingsIdentityProvider, *Response, error)

File: mongodbatlas/federation_settings_connected_org_config.go (hunk 16 above)
Comment (martinstibbe) [reply mongodbatlas/federation_settings_connected_org_config.go]: Updated documentation

File: mongodbatlas/federation_settings_connected_org_config.go (hunk 15 above)
Comment (martinstibbe) [reply mongodbatlas/federation_settings_connected_org_config.go]: Updated documentation

File: mongodbatlas/federation_settings_connected_org_config.go (hunk 14 above)
Comment (martinstibbe) [reply mongodbatlas/federation_settings_connected_org_config.go]: Updated documentation

**Review by gssbzn — APPROVED** [review]
LGTM thanks for all the work here

**Pull request merged**